### PR TITLE
Ports the RBMK from NSV (for real this time)

### PR DESCRIPTION
**Pull request by XeonMations**
## About The Pull Request

Adds in the RBMK nuclear reactor from NSV. (this time hopefully without pull breaking map changes)

All things are able to be ordered from cargo.

Wiki can be found here: https://wiki.beestation13.com/view/Guide_to_the_Nuclear_Reactor

(I accidentally created the wiki on the main wiki instead of a sandbox, I'm sorry)

RBMK Hum noise sourced from [here](https://www.youtube.com/watch?v=KNMU4575bMw), under CC BY 4.0 and edited by me.

## Why It's Good For The Game

More content for engineering!

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

image
image
image
image
image
image

</details>

## Changelog
:cl: XeonMations, Aramix, DungeonVR
add: Added the RBMK reactor, fuel rods and engine sealant, all orderable from cargo!
/:cl:

### CODE REVIEW

**Comment by BlueHNT** [comment]
I fear what the people of bee will do with this. How their antags will inevitably abuse that nuclear fallout. This is going to be good.

**Comment by ClownMoff** [comment]
I think the consequences of blowing up this shit are way to strong, it can fuck entire rounds to a point that is not funny at all

File: code/modules/power/rbmk/_rbmk_defines.dm [hunk 1]
```diff
@@ -31,3 +31,6 @@
 #define REACTOR_EMERGENCY_PERCENT 25
 #define REACTOR_DANGER_PERCENT 50
 #define REACTOR_WARNING_PERCENT 100
+
+//Blame Aramix for this monstrosity. In case you are wondering what this is, it is exp(0.005669), which is used for graphing reactor damage calculations (constant)
+#define REACTOR_DAMAGE_MAGIC_NUMBER 1.0056850991882220077025096652119341723277747452195921845851
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/_rbmk_defines.dm]: as far as I know, Byond can't handle this long shit decimals. some numbers will be lost even, makes it inaccurate.
image

**Comment by MarkusLarsson421** [comment]
Allow items on top of it to be fried. I want my meat RBMK cooked.

File: code/modules/power/rbmk/_rbmk_defines.dm [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#define RBMK_TEMPERATURE_OPERATING 640 //Celsius
```
Comment (PestoVerde322) [inline code/modules/power/rbmk/_rbmk_defines.dm:1]: Curious, but why not use kelvins?

Comment (XeonMations) [reply code/modules/power/rbmk/_rbmk_defines.dm:1]: Helps with displaying correct temperatures, not like the reactor is going to be less than 0 either way

**Comment by PowerfulBacon** [comment]
Not against this being merged, but I don't see the necessity for this feature nor the value that it actually brings.

**Comment by Therealdoooc213** [comment]
> Not against this being merged, but I don't see the necessity for this feature nor the value that it actually brings. We don't want to use this in place of the supermatter crystal which leaves it in a place where its going to be unmaintained and never used.

I don't get this, people aren't going to use a RBMK to replace a roundstart supermatter crystal, it's harder to get and its delamination effects are way worse

**Comment by MarkusLarsson421** [comment]
> We don't want to use this in place of the supermatter crystal which leaves it in a place where its going to be unmaintained and never used.

If it is hidden in some corner of the game, it will be unmaintained because no one will ever go to that corner and know of its existence. I would say we SHOULD allow the SM to be replaced with other engines if the situation desires it. Because then people will actually care for it and its functionality further into the feature. 

With how hidden Derelict Stations are, it is no wonder how unmaintained it is. Not because PRs are being shut down, but because no one knows of its existence to make the PRs to begin with.

**Comment by PowerfulBacon** [comment]
A new mapping standard won't be set for stations to have their SM replaced with RBMK. We will suffer from having to maintain multiple engines, and the RBMK is not an appropriate option for a full SM replacement.

**Comment by Rukofamicom** [comment]
I really love the RBMK, but I ultimately agree with Bacon.

I don't want to replace the supermatter crystals with it, and I don't believe this should be merged unless it is replacing them.

**Comment by Isaacnml** [comment]
Please dont kill it, from initial thought it was supposed to replace the defunct antimatter reactor, which it does a pretty good job, never meant to replace the SM in any way and engineering desperately needs new content to well.. play around gameplay loop currently is really vague

It's a replacement incase the SM explodes, not a full replacement for the SM

**Comment by MarkusLarsson421** [comment]
Almost a year in development and now the close keys are coming in? Jesus Christ. A year of Xeons life wasted.

**Comment by XeonMations** [comment]
> from initial thought it was supposed to replace the defunct antimatter reactor

But yes, the idea from the start was as a replacement for the now gone antimatter reactor.

As of current, if the SM explodes, the station only has solars as a power source unless they want to take a rather long time setting up a black hole (good luck with no power) or a tesla (Again, good luck without power).

Or engineering could just set up a second SM in the form of a shard, but I'm unsure if they'd put that much effort in before the shuttle is called.

**Comment by EvilDragonfiend** [comment]
I am honestly fine with this, but I suppose we might need a kind of mapping tool to replace (or some ingame stuff to make a dedicated area quickly). Constructing a new engine for a whole round was tedious.

**Comment by XeonMations** [comment]
> we might need a kind of mapping tool to replace (or some ingame stuff to make a dedicated area quickly). Constructing a new engine for a whole round was tedious.

Entirely fair, if wished, i am able to add one.

**Comment by EvilDragonfiend** [comment]
> > we might need a kind of mapping tool to replace (or some ingame stuff to make a dedicated area quickly). Constructing a new engine for a whole round was tedious.
> 
> Entirely fair, if wished, i am able to add one.

not right now. we are in a mapping disaster rn.

**Comment by ClownMoff** [comment]
> Not against this being merged, but I don't see the necessity for this feature nor the value that it actually brings. We don't want to use this in place of the supermatter crystal which leaves it in a place where its going to be unmaintained and never used.



> I really love the RBMK, but I ultimately agree with Bacon.
> 
> I don't want to replace the supermatter crystals with it, and I don't believe this should be merged unless it is replacing them.

This just reads as new content is le bad because it is. This isn't gonna replace the SM nor is what the PR attempted to, it's just another option for power. It looks like you guys would key a port of teslas and singulos if they weren't on the game already. Just give people more stuff to do. The server is getting stale. There aren't any big drops of content. We keep nerfing or removing shit and yet whenever anything different from the status quo attempts to get merged it's keyed. Either figure out what do you guys want BeeStation to be already or drop the "bloat" or "unnecesary" argument because while we keep waiting for that big BeeStation overhaul project ™️ it looks like nothing significant gets added to the game, but nerfs and removal keep going on. And this is not only me I'm talking I've seen enough people these last 2 day complaining about the state of the server and losing interest on the server and looking for alternatives, and a lot of complains came from the development sides of things. I will keep being here even though I feel what I do doesn't get appreciated because I like this community a lot, but I for sure will think twice before commiting to big projects if it's just gonna get keyed because "new content is bad because maybe it doesn't get used". I highly suggest you sit down, get your thoughts together as the development team and be trasnparent with the community with what you guys want BeeStation to be and what we can expect. I don't want to open NovemberStation PR after refining the map and seeing Ruko add a key 5 minutes later because urr "the map is not on space so maybe its too different". I want a clear direction or where we are going and what is expected of us. "Nothing is expected of you this is a open sourced project". Well, it doesn't look like it because you clearly have a bias on what to add and what to not add while we wait for this huge BeeStation big change and overhaul. You are on the verge losing old players that have poured heart and soul on this server over constant changes that are just "makes the game feel more tedious to play" while nothign new or refreshing gets added. There is no incentive from the development team to recruit new talent, no big projects that we can work on, no events, no nothing. Silence and waiting days, weeks or months for a review (I really understand we lack the manpower in maintenance) and then who knows, maybe a key because your new PR didn't allign with what we wanted to do.

**Review by PowerfulBacon — APPROVED** [review]
Eh whatever, this can replace the antimatter generator.

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
Replace arbitrary names like R or N. It tells no context.

File: code/modules/power/rbmk/rbmk_monitoring.dm [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+		return FALSE
+	var/new_status = get_status()
+	if(last_status != new_status)
+		last_status = new_status
+		ui_header = "smmon_[last_status].gif"
+		program_icon_state = "smmon_[last_status]"
+		if(istype(computer))
+			computer.update_icon()
+	if(world.time >= next_stat_interval)
+		next_stat_interval = world.time + 1 SECONDS //You only get a slow tick.
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_monitoring.dm]: this can be cooldown macro.

File: code/modules/power/rbmk/rbmk_monitoring.dm [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+		if(kpaData.len > 100) //Only lets you track over a certain timeframe.
+			kpaData.Cut(1, 2)
+		powerData += (reactor) ? reactor.power*10 : 0 //We scale up the figure for a consistent:tm: scale
+		if(powerData.len > 100) //Only lets you track over a certain timeframe.
+			powerData.Cut(1, 2)
+		tempInputData += (reactor) ? reactor.last_coolant_temperature : 0 //We scale up the figure for a consistent:tm: scale
+		if(tempInputData.len > 100) //Only lets you track over a certain timeframe.
+			tempInputData.Cut(1, 2)
+		tempOutputdata += (reactor) ? reactor.last_output_temperature : 0 //We scale up the figure for a consistent:tm: scale
+		if(tempOutputdata.len > 100) //Only lets you track over a certain timeframe.
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_monitoring.dm]: please replace len to length()

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 5]
```diff
@@ -0,0 +1,613 @@
+//This section contain all procs that helps building, destroy and control the RBMK
+
+/obj/machinery/atmospherics/components/unary/rbmk/core/attackby(obj/item/attacked_item, mob/user, params)
+	if(istype(attacked_item, /obj/item/fuel_rod))
+		if(power >= 20)
+			to_chat(user, "<span class='notice'>You cannot insert fuel into [src] when it has been raised above 20% power.</span>")
+			return FALSE
+		if(fuel_rods.len >= 5)
+			to_chat(user, "<span class='warning'>[src] is already at maximum fuel load.</span>")
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]: this len to length too please

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 6]
```diff
@@ -0,0 +1,613 @@
...
+		to_chat(user, "<span class='notice'>You start to insert [attacked_item] into [src]...</span>")
+		radiation_pulse(src, temperature) //Wear protective equipment when even breathing near a reactor!
+		if(do_after(user, 5 SECONDS, target=src))
+			if(!length(fuel_rods))
+				activate(user) //That was the first fuel rod. Let's heat it up.
+			fuel_rods += attacked_item
+			attacked_item.forceMove(src)
+			if(!(length(fuel_rods) == 1)) // Not the first fuel rod? Play the sound.
+				playsound(src, pick('sound/effects/rbmk/switch1.ogg','sound/effects/rbmk/switch2.ogg','sound/effects/rbmk/switch3.ogg'), 100, FALSE)
+			update_appearance()
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
			if(length(fuel_rods) == 0)
				activate(user) //That was the first fuel rod. Let's heat it up.
			else // Not the first fuel rod? Play the sound.
				playsound(src, pick('sound/effects/rbmk/switch1.ogg','sound/effects/rbmk/switch2.ogg','sound/effects/rbmk/switch3.ogg'), 100, FALSE)
			fuel_rods += attacked_item
			attacked_item.forceMove(src)
			update_appearance()
```

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+			to_chat(user, "<span class='notice'>You cannot repair [src] while it is running at above 20% power.</span>")
+			return FALSE
+		if(critical_threshold_proximity <= 0.875*critical_threshold_proximity_archived)
+			to_chat(user, "<span class='notice'>[src]'s seals are already in-tact, repairing them further would require a new set of seals.</span>")
+			return FALSE
+		if(critical_threshold_proximity >= 0.5 * critical_threshold_proximity_archived) //Heavily damaged.
+			to_chat(user, "<span class='notice'>[src]'s reactor vessel is cracked and worn, you need to repair the cracks with a welder before you can repair the seals.</span>")
+			return FALSE
+		if(do_after(user, 5 SECONDS, target=src))
+			if(critical_threshold_proximity <= 0.875*critical_threshold_proximity_archived)	//They might've stacked doafters
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]: where did this `0.875` come from?
This should be a define macro.

also, why `0.5` too?

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+				to_chat(user, "<span class='notice'>[src]'s seals are already in-tact, repairing them further would require a new set of seals.</span>")
+				return FALSE
+			playsound(src, 'sound/effects/spray2.ogg', 50, 1, -6)
+			user.visible_message("<span class='warning'>[user] applies sealant to some of [src]'s worn out seals.</span>", "<span class='notice'>You apply sealant to some of [src]'s worn out seals.</span>")
+			critical_threshold_proximity -= 10
+			critical_threshold_proximity = clamp(critical_threshold_proximity, 0, initial(critical_threshold_proximity))
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
			critical_threshold_proximity = clamp(critical_threshold_proximity - 10, 0, initial(critical_threshold_proximity))
```

and this 10 should be a macro. I know what this number means, but other code readers might not.
`#define SEALNANT_REPAIR_VALUE 10` like this?
or do `/sealant /var/repair_power = 10`, and `critical_threshold_proximity - sealant.repair_power`

File: code/modules/power/rbmk/rbmk_procs.dm
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm:1]: at `power >= 20`, this 20 should be `#define SAFE_POWER_LEVEL 20` smh. I am not sure what's better name.

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 9]
```diff
@@ -0,0 +1,613 @@
...
+	var/obj/item/held_item = user.get_active_held_item()
+	if(!issilicon(user) && held_item?.GetComponent(/datum/component/buffer))
+		return held_item?.GetComponent(/datum/component/buffer)
+	else if(isAI(user))
+		var/mob/living/silicon/ai/ai_user = user
+		return ai_user.aiMulti.GetComponent(/datum/component/buffer)
+	else if(iscyborg(user) && in_range(user, src))
+		if(held_item?.GetComponent(/datum/component/buffer))
+			return held_item?.GetComponent(/datum/component/buffer)
+	return null
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
/obj/machinery/atmospherics/components/unary/rbmk/proc/get_held_buffer_item(mob/user)
	if(isAI(user))
		var/mob/living/silicon/ai/ai_user = user
		return ai_user.aiMulti.GetComponent(/datum/component/buffer)

	var/obj/item/held_item = user.get_active_held_item()
	var/found_component = held_item?.GetComponent(/datum/component/buffer)
	if(found_component && in_range(user, src))
		return found_component

	return nulll
```
If I understand this correctly, It should check `in_range` regardless of being cyborg or carbon whatever right?

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 10]
```diff
@@ -0,0 +1,613 @@
...
+					/obj/item/RBMK_box/body/waste_output,
+					/obj/item/RBMK_box/body,
+					/obj/item/RBMK_box/body,
+					/obj/item/RBMK_box/body,
+					/obj/item/RBMK_box/body,
+					/obj/item/RBMK_box/body)
+	for(var/item in parts)
+		new item(get_turf(src))
+	I.play_tool_sound(src, 50)
+	QDEL_NULL(src)
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]: it's just `qdel`

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 11]
```diff
@@ -0,0 +1,613 @@
...
+
+/obj/machinery/atmospherics/components/unary/rbmk/core/proc/has_fuel()
+	return length(fuel_rods)
+
+
+/obj/machinery/atmospherics/components/unary/rbmk/core/proc/removeFuelRod(mob/user, /obj/machinery/atmospherics/components/unary/rbmk/core/reactor)
+	if(src.power > 20)
+		to_chat(user, "<span class='warning'>You cannot remove fuel from [src] when it is above 20% power.</span>")
+		return FALSE
+	if(!length(fuel_rods))
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
	if(length(fuel_rods) == 0)
```
This reads better, I guess? nitpick.

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 12]
```diff
@@ -0,0 +1,613 @@
...
+		use_power((power + 1) * IDLE_POWER_USE)
+	return TRUE
+
+/obj/machinery/atmospherics/components/unary/rbmk/core/proc/has_fuel()
+	return length(fuel_rods)
+
+
+/obj/machinery/atmospherics/components/unary/rbmk/core/proc/removeFuelRod(mob/user, /obj/machinery/atmospherics/components/unary/rbmk/core/reactor)
+	if(src.power > 20)
+		to_chat(user, "<span class='warning'>You cannot remove fuel from [src] when it is above 20% power.</span>")
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]: reminder that this 20 needs a macro.

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 13]
```diff
@@ -0,0 +1,613 @@
...
+/obj/machinery/atmospherics/components/unary/rbmk/core/proc/get_status()
+	var/integrity = get_integrity_percent()
+	if(integrity < REACTOR_MELTING_PERCENT)
+		return REACTOR_MELTING
+	if(integrity < REACTOR_EMERGENCY_PERCENT)
+		return REACTOR_EMERGENCY
+	if(integrity < REACTOR_DANGER_PERCENT)
+		return REACTOR_DANGER
+	if(integrity < REACTOR_WARNING_PERCENT)
+		return REACTOR_NOMINAL
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
	switch(get_integrity_percent())
		if(0 to REACTOR_MELTING_PERCENT)
			return REACTOR_MELTING
		if(REACTOR_MELTING_PERCENT to REACTOR_EMERGENCY_PERCENT)
			return REACTOR_EMERGENCY
		if(REACTOR_EMERGENCY_PERCENT to REACTOR_DANGER_PERCENT)
			return REACTOR_DANGER
		if(REACTOR_DANGER_PERCENT to REACTOR_WARNING_PERCENT)
			return REACTOR_NOMINAL
```
using switch is better when it is 0 to 100 condition.

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 14]
```diff
@@ -0,0 +1,613 @@
...
+/obj/machinery/atmospherics/components/unary/rbmk/core/proc/meltdown()
+	set waitfor = FALSE
+	SSair.atmos_machinery -= src //Annd we're now just a useless brick.
+	update_icon()
+	STOP_PROCESSING(SSmachines, src)
+	AddComponent(/datum/component/radioactive, 15000 , src)
+	var/turf/reactor_turf = get_turf(src)
+	var/rbmkzlevel = reactor_turf.get_virtual_z_level()
+	for(var/mob/player_mob in GLOB.player_list)
+		if(player_mob.get_virtual_z_level() == rbmkzlevel)
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
		if(compare_z(rbmkzlevel, player_mob.get_virtual_z_level())
```
This is a proper way to compare Zs. getting z value only tells the number, not checking if they're in the same z category. (i.e. station)

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 15]
```diff
@@ -0,0 +1,613 @@
...
+			to_chat(player_mob, "<span class='userdanger'>You hear a horrible metallic hissing.</span>")
+			SEND_SIGNAL(player_mob, COMSIG_ADD_MOOD_EVENT, "delam", /datum/mood_event/delam) //Might as well use the same moodlet since its essentialy the same thing happening
+
+	for(var/obj/machinery/power/apc/apc in GLOB.apcs_list)
+		if(src.get_virtual_z_level() == apc.get_virtual_z_level() && prob(70))
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
		if(prob(70) && compare_z(rbmkzlevel, apc.get_virtual_z_level())
```
reusing the variable `rbmkzlevel`.

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 16]
```diff
@@ -0,0 +1,613 @@
...
+	SSblackbox.record_feedback("tally", "engine_stats", 1, "failed")
+	SSblackbox.record_feedback("tally", "engine_stats", 1, "agcnr")
+	Destroy()
+
+/obj/machinery/atmospherics/components/unary/rbmk/core/proc/blowout()
+	explosion(get_turf(src), GLOB.MAX_EX_DEVESTATION_RANGE, GLOB.MAX_EX_HEAVY_RANGE, GLOB.MAX_EX_LIGHT_RANGE, GLOB.MAX_EX_FLASH_RANGE)
+	var/turf/reactor_turf = get_turf(src)
+	var/rbmkzlevel = reactor_turf.get_virtual_z_level()
+	for(var/mob/player_mob in GLOB.player_list)
+		if(player_mob.get_virtual_z_level() == rbmkzlevel)
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
		if(compare_z(rbmkzlevel, player_mob.get_virtual_z_level())
```

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 17]
```diff
@@ -0,0 +1,613 @@
...
+			SEND_SOUND(player_mob, 'sound/effects/rbmk/explode.ogg')
+			to_chat(player_mob, "<span class='userdanger'>You hear a horrible metallic explosion.</span>")
+			SEND_SIGNAL(player_mob, COMSIG_ADD_MOOD_EVENT, "delam", /datum/mood_event/delam) //Might as well use the same moodlet since its essentialy the same thing happening
+	for(var/nuclear_sludge_landmark in GLOB.landmarks_list)
+		if(istype(nuclear_sludge_landmark, /obj/modules/power/rbmk/nuclear_sludge_spawner))
+			var/obj/modules/power/rbmk/nuclear_sludge_spawner/nuclear_sludge_spawner = nuclear_sludge_landmark
+			if(src.get_virtual_z_level() == nuclear_sludge_spawner.get_virtual_z_level()) //Begin the SLUDGING
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
			if(compare_z(rbmkzlevel, nuclear_sludge_spawner.get_virtual_z_level()) //Begin the SLUDGING
```

File: code/modules/power/rbmk/rbmk_monitoring.dm [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+			tempOutputdata.Cut(1, 2)
+
+/datum/computer_file/program/nuclear_monitor/on_start(mob/living/user)
+	. = ..(user)
+	//No reactor? Go find one then.
+	if(!reactor)
+		for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
+			if(user.get_virtual_z_level() == reactor_core.get_virtual_z_level())
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_monitoring.dm]:
```suggestion
		var/core_z_level = reactor_core.get_virtual_z_level()
		for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
			if(compare_z(core_z_level , user.get_virtual_z_level())
```

File: code/modules/power/rbmk/rbmk_monitoring.dm [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+
+/datum/computer_file/program/nuclear_monitor/ui_act(action, params)
+	if(..())
+		return TRUE
+
+	switch(action)
+		if("swap_reactor")
+			var/list/choices = list()
+			for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
+				if(usr.get_virtual_z_level() != reactor_core.get_virtual_z_level())
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_monitoring.dm]:
```suggestion
			var/user_z_level = usr.get_virtual_z_level()
			for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
				if(!compare_z(user_z_level, reactor_core.get_virtual_z_level())
```
I am not sure if doing `usr` is right here, but whatever. it won't be breaking the game...

File: code/modules/power/rbmk/rbmk_parts.dm [hunk 20]
```diff
@@ -0,0 +1,257 @@
...
+	. = ..()
+	attack_hand(user)
+
+/obj/machinery/computer/reactor/attack_ai(mob/user)
+	. = ..()
+	attack_hand(user)
+
+/obj/machinery/computer/reactor/attackby(obj/item/tool, mob/user, params)
+	if(tool.tool_behaviour == TOOL_MULTITOOL)
+		var/datum/component/buffer/heldmultitool = get_held_buffer_item(usr)
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_parts.dm]:
```suggestion
		var/datum/component/buffer/heldmultitool = get_held_buffer_item(user)
```
do not use `usr` when you can have `user`

File: code/modules/power/rbmk/rbmk_parts.dm [hunk 21]
```diff
@@ -0,0 +1,257 @@
...
+	var/obj/item/held_item = user.get_active_held_item()
+	if(!issilicon(user) && held_item?.GetComponent(/datum/component/buffer))
+		return held_item?.GetComponent(/datum/component/buffer)
+	else if(isAI(user))
+		var/mob/living/silicon/ai/U = user
+		return U.aiMulti.GetComponent(/datum/component/buffer)
+	else if(iscyborg(user) && in_range(user, src))
+		if(held_item?.GetComponent(/datum/component/buffer))
+			return held_item?.GetComponent(/datum/component/buffer)
+	return null
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_parts.dm]:
```suggestion
/obj/machinery/computer/reactor/proc/get_held_buffer_item(mob/user)
	if(isAI(user))
		var/mob/living/silicon/ai/ai_user = user
		return ai_user.aiMulti.GetComponent(/datum/component/buffer)

	var/obj/item/held_item = user.get_active_held_item()
	var/found_component = held_item?.GetComponent(/datum/component/buffer)
	if(found_component && in_range(user, src))
		return found_component

	return nulll
```

File: code/modules/power/rbmk/fuel_rods.dm [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+	AddComponent(/datum/component/radioactive, rad_strength, src) // This should be temporary for it won't make rads go lower than 350
+	if(process)
+		START_PROCESSING(SSobj, src)
+
+/obj/item/fuel_rod/Destroy()
+	if(process)
+		STOP_PROCESSING(SSobj, src)
+	var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core = loc
+	if(istype(reactor_core))
+		reactor_core.fuel_rods -= src
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/fuel_rods.dm]:
```suggestion
	if(istype(loc, /obj/machinery/atmospherics/components/unary/rbmk/core))
			var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core = loc
		reactor_core.fuel_rods -= src
```

File: code/modules/power/rbmk/fuel_rods.dm [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+	return ..()
+
+// This proc will try to convert your fuel rod if you don't override this proc
+// So, ideally, you should write an override of this for every fuel rod you want to create
+/obj/item/fuel_rod/proc/depletion_final(result_rod)
+	if(!result_rod)
+		return
+	var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core = loc
+	// Rod conversion is moot when you can't find the reactor
+	if(istype(reactor_core))
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/fuel_rods.dm]:
```suggestion
	// Rod conversion is moot when you can't find the reactor
	if(istype(loc, /obj/machinery/atmospherics/components/unary/rbmk/core))
		var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core = loc
```

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
Code looks alright as long as change requests are applied.

File: code/modules/power/rbmk/rbmk_monitoring.dm [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+		if(length(tempOutputdata) > 100) //Only lets you track over a certain timeframe.
+			tempOutputdata.Cut(1, 2)
+
+/datum/computer_file/program/nuclear_monitor/on_start(mob/living/user)
+	. = ..(user)
+	//No reactor? Go find one then.
+	if(!reactor)
+		for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
+			var/core_z_level = reactor_core.get_virtual_z_level()
+			if(compare_z(core_z_level , user.get_virtual_z_level()))
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_monitoring.dm]:
```suggestion
		var/user_z_level = user.get_virtual_z_level())
		for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
			if(compare_z(reactor_core.get_virtual_z_level(), user_z_level))
```
Sorry, my mistake. It's reversed.

File: code/modules/power/rbmk/rbmk_monitoring.dm [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+		return TRUE
+
+	switch(action)
+		if("swap_reactor")
+			var/list/choices = list()
+			for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
+				var/core_z_level = reactor_core.get_virtual_z_level()
+				if(!compare_z(core_z_level , usr.get_virtual_z_level()))
+					continue
+				choices += reactor_core
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_monitoring.dm]:
```suggestion
			var/user_z_level = usr.get_virtual_z_level()
			for(var/obj/machinery/atmospherics/components/unary/rbmk/core/reactor_core in GLOB.machines)
				if(!compare_z(reactor_core.get_virtual_z_level(), user_z_level))
					continue
				choices += reactor_core
```
Same mistake here too

File: code/modules/power/rbmk/rbmk_procs.dm [hunk 26]
```diff
@@ -0,0 +1,611 @@
...
+			to_chat(user, "<span class='notice'>You cannot insert fuel into [src] when it has been raised above [SAFE_POWER_LEVEL]% power.</span>")
+			return FALSE
+		if(length(fuel_rods) >= 5)
+			to_chat(user, "<span class='warning'>[src] is already at maximum fuel load.</span>")
+			return FALSE
+		to_chat(user, "<span class='notice'>You start to insert [attacked_item] into [src]...</span>")
+		radiation_pulse(src, temperature) //Wear protective equipment when even breathing near a reactor!
+		if(do_after(user, 5 SECONDS, target=src))
+			if(length(fuel_rods) == 0)
+				activate(user) //That was the first fuel rod. Let's heat it up.
```
Comment (EvilDragonfiend) [inline code/modules/power/rbmk/rbmk_procs.dm]:
```suggestion
			if(length(fuel_rods) >= 5)
				to_chat(user, "<span class='warning'>[src] is already at maximum fuel load.</span>")
				return FALSE
			else if(length(fuel_rods) == 0)
				activate(user) //That was the first fuel rod. Let's heat it up.
```
If multiple people trying this, It's possible to have +5 rods.

**Review by EvilDragonfiend — CHANGES_REQUESTED** [review]
Would be a final.

**Review by EvilDragonfiend — APPROVED** [review]
okay, we'll see how this goes.

**Comment by BarteG44** [comment]
YES YES YES YES

**Pull request merged**